design/healcheck-endpoint.md

```diff
@@ -0,0 +1,159 @@
+or integrating with Microprofile Health Subsystem. An additional problem is that Wildfly provides its own
+`HealthReporter` class (see [the code](https://github.com/wildfly/wildfly/blob/master/microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/HealthReporter.java))
+that doesn't use groups. The Grouping functionality is essential for this design (explained below).
+
+The biggest downside of providing Keycloak's own implementation is that we'll end up with 2 health endpoints -
+one for Wildfly (`http://localhost:9990/health`) and one for Keycloak (`https://localhost:8443/auth/health`). We
+need to be very clear in the documentation that the latter is the proper one.
+
+Having the above in mind, the design assumes integrating the new Healthcheck API with Keycloak's REST interface.
+The implementation will create additional `/health` endpoint that will parse incoming query parameters. Once all
+parameters are parsed, the implementation will extract all additional checks specified in `fail-on` parameter.
+Those checks will be reflected as groups in `KeycloakHealthReporter`. The endpoint implementation should be similar
+to Smallrye's `SmallRyeHealthGroupServlet` (see [the code](https://github.com/smallrye/smallrye-health/blob/811c63198e5876704cebf3691fe3e4dba499fa8b/testsuite/tck/src/main/java/io/smallrye/health/SmallRyeHealthGroupServlet.java#L13)).
+
+#### Healthcheck API for Liveness and Readiness probes
+
+The new Healthcheck API gives a lot of flexibility for defining both Liveness and Readiness Checks. The proposed usage
+looks like the following:
+
+    # Liveness Check
+    curl https://<keycloak-url>:8443/auth/health
+
+    # The most basic Readiness Check
+    curl https://<keycloak-url>:8443/auth/health?fail-on=db
+
+    # The more advanced Readiness Check
+    curl https://<keycloak-url>:8443/auth/health?fail-on=db,idp-google,user-federation-myldap
+
+The REST response has been defined in [Microprofile Health Spec](https://download.eclipse.org/microprofile/microprofile-health-3.0-RC4/microprofile-health-spec.html#_appendix_a_rest_interfaces_specifications).
+There are 3 possible HTTP Status codes that might be returned from the endpoint:
+* 200 - Healthcheck passes
+* 500, 503 - Healthcheck fails
+
+The REST response body contains additional information helpful for debugging. A non-ready response call for
+`https://<keycloak-url>:8443/auth/health?fail-on=db` might look like the following:
+
+```json
+{
+  "status": "DOWN",
+  "checks": [
+    {
+      "name": "org.keycloak.health.DatabaseCheck",
+      "status": "DOWN",
+      "data": {
+        "cause": "failing for 30s"
+      }
+    }
+  ]
+}
+```
+
+## The workaround for the database upgrade
+
+Since Undertow subsystem starts after the database migration, there's no way provide a working REST Health Endpoint
```

Comment:
Can't you just have a database migration service separated from the keycloak other elements? Greenlighting the start?

Reply:
Unfortunately I don't know the history behind the current implementation. Perhaps @stianst @hmlnarik or @mposolda could chime in and give us some more context. I'm not sure if this is achievable or not.

Reply:
We're going to need a different strategy for DB upgrades (both before and after we support zero downtime upgrades). Not sure what is generally best practices here on K8s.

Reply:
I believe the key requirement is that Liveness Probe needs to be up and Readiness Probe down. This workaround assumes this behavior. 

But we can change the approach to whenever we want. At the moment, Keycloak Operator uses a ConfigMap with a raw Liveness/Readiness checks (see [the code](https://github.com/keycloak/keycloak-operator/blob/master/pkg/model/keycloak_probes.go)). We can put whatever we want there.

Reply:
I added small clarification about the Readiness Probe.

Let me resolve this conversation than.